dg.go

```diff
@@ -1,315 +1,480 @@
+// dependency to the `to` node.
+func (d *directedGraph[NodeType]) connectNodes(fromID, toID string, dependencyType DependencyType) error {
+	d.lock.Lock()
+	defer d.lock.Unlock()
+	// Make sure both nodes exist and are not deleted.
+	fromNode, ok := d.nodes[fromID]
+	if !ok {
+		return &ErrNodeNotFound{fromID}
+	} else if fromNode.deleted {
+		return &ErrNodeDeleted{fromID}
+	}
+	toNode, ok := d.nodes[toID]
+	if !ok {
+		return &ErrNodeNotFound{toID}
+	} else if toNode.deleted {
+		return &ErrNodeDeleted{toID}
+	}
+	// Check that it's a non-self and non-duplicate connection.
+	if fromID == toID {
+		return &ErrCannotConnectToSelf{fromID}
+	}
+	if _, ok := d.connectionsFromNode[fromID][toID]; ok {
+		return &ErrConnectionAlreadyExists{fromID, toID}
+	}
+	// Update the mappings.
+	d.connectionsFromNode[fromID][toID] = struct{}{}
+	d.connectionsToNode[toID][fromID] = struct{}{}
+	// Update the dependencies
+	toNode.outstandingDependencies[fromID] = dependencyType
+	return nil
+}
+
+func (d *directedGraph[NodeType]) PushStartingNodes() error {
+	d.lock.Lock()
+	defer d.lock.Unlock()
+
+	for nodeID, n := range d.nodes {
+		if len(n.outstandingDependencies) == 0 {
+			d.readyForProcessing[nodeID] = n
+		}
+	}
+	return nil
+}
+
+func (d *directedGraph[NodeType]) PopReadyNodes() []*node[NodeType] {
+	d.lock.Lock()
+	// Transfer the map to a local variable to minimize time locked, and reset the graph's value.
+	readyMap := d.readyForProcessing
+	d.readyForProcessing = make(map[string]*node[NodeType])
+	d.lock.Unlock()
+	// Copy the map nodes to a slice to comply with the interface.
+	result := make([]*node[NodeType], len(readyMap))
+	i := 0
+	for _, node := range readyMap {
+		result[i] = node
+		i += 1
+	}
+	return result
 }
 
 type node[NodeType any] struct {
-    deleted bool
-    id      string
-    item    NodeType
-    dg      *directedGraph[NodeType]
+	deleted                 bool
+	id                      string
+	item                    NodeType
+	ready                   bool
+	status                  ResolutionStatus
+	outstandingDependencies map[string]DependencyType
+	dg                      *directedGraph[NodeType]
+}
+
+func (n *node[NodeType]) ID() string {
+	return n.id
+}
+
+func (n *node[NodeType]) Item() NodeType {
+	return n.item
+}
+
+func (n *node[NodeType]) ResolutionStatus() ResolutionStatus {
+	n.dg.lock.Lock()
+	defer n.dg.lock.Unlock()
+	return n.status
+}
+
+func (n *node[NodeType]) IsReady() bool {
+	n.dg.lock.Lock()
+	defer n.dg.lock.Unlock()
+	return n.ready
+}
+
+func (n *node[NodeType]) OutstandingDependencies() map[string]DependencyType {
+	n.dg.lock.Lock()
+	defer n.dg.lock.Unlock()
+	return maps.Clone(n.outstandingDependencies)
+}
+
+func (n *node[NodeType]) ResolveNode(status ResolutionStatus) error {
+	n.dg.lock.Lock()
+	defer n.dg.lock.Unlock()
+	return n.resolveNode(status)
+}
+
+func (n *node[NodeType]) resolveNode(status ResolutionStatus) error {
```

Comment:
Dang, I had to look back and forth twice before I realized this wasn't simply a redeclaration of the exported `ResolveNode`. A comment or two would help a lot... and a different name would be even better.